include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -0,0 +1,303 @@
+        boost::math::detail::IntervalSieve sieve(current_lower_bound, upper_bound, primes, prime_vectors.back());
+    }));
+
+    for(auto &&future : future_manager)
+    {
+        if(future.valid())
+        {
+            future.get();
+        }
+    }
+
+    for(auto &v : prime_vectors)
+    {
+        resultant_primes.insert(resultant_primes.end(), v.begin(), v.end());
+    }
+}
+
+template<class Integer, class Container>
+void segmented_sieve(Integer lower_bound, Integer upper_bound, Container &resultant_primes)
+{
+    Integer limit{static_cast<Integer>(std::floor(std::sqrt(static_cast<double>(upper_bound)))) + 1};
+    std::vector<Integer> primes {};
+    primes.reserve(static_cast<double>(limit) / std::log(static_cast<double>(limit)));
+
+    // Prepare for max value so you do not have to calculate this again
+    if(limit < 4096)
+    {
+        boost::math::detail::linear_sieve(static_cast<Integer>(limit), primes);
+    }
+
+    else
+    {
+        //boost::math::detail::mask_sieve(static_cast<Integer>(2), limit, primes);
+        boost::math::detail::linear_sieve(static_cast<Integer>(4096), primes);
+        boost::math::detail::segmented_sieve(static_cast<Integer>(4096), limit, primes, primes);
+    }
+
+    boost::math::detail::segmented_sieve(lower_bound, upper_bound, primes, resultant_primes);
+}
+
+template<class Integer, class Container>
+void sequential_segmented_sieve(Integer lower_bound, Integer upper_bound, Container &resultant_primes)
+{
+    const Integer L1_SIZE {32768};
+    const Integer interval {L1_SIZE * 8};
+    Integer current_lower_bound{lower_bound};
+    Integer current_upper_bound{current_lower_bound + interval};
+
+    if(current_upper_bound > upper_bound)
+    {
+        current_upper_bound = upper_bound;
+    }
+
+    std::size_t ranges {static_cast<std::size_t>((upper_bound - lower_bound) / interval)};
+
+    boost::math::detail::IntervalSieve sieve(current_lower_bound, current_upper_bound, resultant_primes, resultant_primes);
+    if(ranges == 0)
+    {
+        return;
+    }
+
+    for(std::size_t i {}; i < ranges; ++i)
+    {
+        current_lower_bound = current_upper_bound;
+        current_upper_bound += interval;
+        if(current_upper_bound > upper_bound)
+        {
+            current_upper_bound = upper_bound;
+        }
+        sieve.NewRange(current_lower_bound, current_upper_bound, resultant_primes);
+    }
+}
+} // End namespace detail
+
+template<class Integer>
+constexpr void prime_reserve(Integer upper_bound, std::vector<Integer> &prime_container)
+{
+    prime_container.reserve(static_cast<double>(upper_bound) / std::log(static_cast<double>(upper_bound)));
+}
+
+template<class ExecutionPolicy, class Integer, class Container>
+void prime_sieve(ExecutionPolicy&& policy, Integer upper_bound, Container &primes)
+{
+    if(upper_bound == 2)
+    {
+        return;
+    }
+
+    if(upper_bound <= 4096)
```

Comment:
4096 is a magic number here that's likely to change between CPU architectures. 
I would make it a parameter and either i) for the least effort, just make 4096 the default value, or ii) write a wrapper function that uses whatever is available (cache hierarchy, sizes, etc) to tune the value.

Reply:
There are APIs to detect L1 cache size; for instance `perf` knows it.

How is this done? I feel slightly ashamed I have no clue how to do it after this many years programming.

Reply:
@NAThompson On linux you can call [sysconf](https://www.gnu.org/software/libc/manual/html_node/Constants-for-Sysconf.html). How to do it in a portable way? Good question.

Reply:
I'm not aware of a portable way to do it in C++ other than [these C++17 functions](https://en.cppreference.com/w/cpp/thread/hardware_destructive_interference_size) in the thread support library.
I thought there were more, but I can't find them.

Reply:
Maybe Boost has or should have a cross-platform wrapper for that kind of thing?